sentry-spring-boot-starter/src/main/java/io/sentry/spring/boot/SentryAutoConfiguration.java

```diff
@@ -0,0 +1,108 @@
+package io.sentry.spring.boot;
+
+import com.jakewharton.nopen.annotation.Open;
+import io.sentry.core.EventProcessor;
+import io.sentry.core.HubAdapter;
+import io.sentry.core.IHub;
+import io.sentry.core.Integration;
+import io.sentry.core.Sentry;
+import io.sentry.core.SentryClient;
+import io.sentry.core.SentryOptions;
+import io.sentry.core.protocol.SdkVersion;
+import io.sentry.core.transport.ITransport;
+import io.sentry.core.transport.ITransportGate;
+import org.jetbrains.annotations.NotNull;
+import org.springframework.beans.factory.ObjectProvider;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
+import org.springframework.boot.autoconfigure.condition.ConditionalOnWebApplication;
+import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.web.servlet.FilterRegistrationBean;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+
+@Configuration
+@ConditionalOnProperty(name = "sentry.enabled", havingValue = "true", matchIfMissing = true)
+@Open
+public class SentryAutoConfiguration {
+
+  /** Registers general purpose Sentry related beans. */
+  @Configuration
+  @ConditionalOnProperty("sentry.dsn")
+  @EnableConfigurationProperties(SentryProperties.class)
+  @Open
+  static class HubConfiguration {
+
+    @Bean
+    @ConditionalOnMissingBean
+    public Sentry.OptionsConfiguration<SentryOptions> optionsOptionsConfiguration(
+      ObjectProvider<SentryOptions.BeforeSendCallback> beforeSendCallback,
+      ObjectProvider<SentryOptions.BeforeBreadcrumbCallback> beforeBreadcrumbCallback,
+      ObjectProvider<EventProcessor> eventProcessors,
+      ObjectProvider<Integration> integrations,
+      ObjectProvider<ITransportGate> transportGate,
+      ObjectProvider<ITransport> transport) {
+      return options -> {
+        beforeSendCallback.ifAvailable(options::setBeforeSend);
+        beforeBreadcrumbCallback.ifAvailable(options::setBeforeBreadcrumb);
+        eventProcessors.stream().forEach(options::addEventProcessor);
+        integrations.stream().forEach(options::addIntegration);
+        transportGate.ifAvailable(options::setTransportGate);
+        transport.ifAvailable(options::setTransport);
+      };
+    }
+
+    @Bean
+    IHub sentryHub(
+        Sentry.OptionsConfiguration<SentryOptions> optionsConfiguration,
+        SentryProperties properties,
+        ObjectProvider<SentryClient> sentryClient) {
+      Sentry.init(
+          options -> {
+            properties.applyTo(options);
+            optionsConfiguration.configure(options);
+            options.setSentryClientName(BuildConfig.SENTRY_SPRING_BOOT_SDK_NAME);
+            options.setSdkVersion(createSdkVersion(options));
+          });
+      sentryClient.ifAvailable(Sentry::bindClient);
```

Comment:
Unsure about this. The options passed to `init` is used to initialize the client.
What is the use case to take both options and client in here?

Also, binding the client after calling init has some down sides (things we can do during initialization that require a client already, created in the process of init).

If letting the client com from outside will be a must, I suggest we introduce a `ClientFactory` in the options and hook here so that the SDK under the hood (through init) calls that instead of binding after the fact.